src/highlights/highlight_grid.ts

```diff
@@ -1,152 +1,252 @@
+        const prevCells = this.lineCells[line] ?? [];
+        // Fill in the missing cells
+        if (prevCells.length < vimCol) {
+            const missingCells = vimCol - prevCells.length;
+            for (let i = 0; i < missingCells; i++) {
+                prevCells.push({ text: " ", hlId: 0 });
+            }
+        }
 
-    private static splitter = new GraphemeSplitter();
+        const redrawCells: LineCell[] = [];
+        {
+            let currHlId = 0;
+            for (const [text, hlId, repeat] of cells) {
+                if (hlId != null) currHlId = this.groupStore.normalizeHighlightId(hlId);
+                for (let i = 0; i < (repeat ?? 1); i++) {
+                    redrawCells.push({ text, hlId: currHlId });
+                }
+            }
+        }
+        const leftCells = prevCells.slice(0, vimCol);
+        const rightCells = prevCells.slice(vimCol + redrawCells.length);
 
-    constructor() {
-        this.grid = [];
+        this.lineCells[line] = [...leftCells, ...redrawCells, ...rightCells];
     }
 
-    /**
-     * Checks whether or not the contents of a text cell should actually span two cells. For instance:
-     * a  length:1 width:1
-     * 你 length:1 width:2
-     * 🚀 length:2 width:2
-     * 🕵️ length:3 width:2
-     * ❤️ length:2 width:2
-     *
-     * @param cellText The cell text to check
-     * @returns Whether or not this cell is a double-width cell
-     */
-    static isDouble(cellText: string): boolean {
-        return wcswidth(cellText) === 2 || cellText.length > 1;
+    scroll(by: number) {
+        const lineNumbers: number[] = [];
+        this.lineCells.forEach((_, i) => lineNumbers.push(i));
+
+        if (by > 0) {
+            for (const line of lineNumbers) {
+                this.lineCells[line - by] = this.lineCells[line];
+            }
+        } else if (by < 0) {
+            const scrollBy = Math.abs(by);
+            for (const line of lineNumbers.reverse()) {
+                this.lineCells[line + scrollBy] = this.lineCells[line];
+            }
+        }
     }
 
-    /**
-     * Get the width of a piece of text, in cells. Note that the use of tabs is
-     * position dependent, and this function must assume a line with tabs starts
-     * at position 0in the editor.
-     *
-     * @param text The text to get the width of
-     * @param tabSize The size of tab-widths
-     * @returns The width of the text, in cells
-     */
-    static getWidth(text: string, tabSize: number): number {
-        const t = expandTabs(text, tabSize);
-        return HighlightGrid.splitGraphemes(t).reduce((p, c) => p + (HighlightGrid.isDouble(c) ? 2 : 1), 0);
+    // #endregion
+
+    // #region Calculate Decorations
+
+    // decoration type -> decoration options
+    public getDecorations(
+        editor: TextEditor,
+        startLine: number,
+        endLine: number,
+    ): Map<TextEditorDecorationType, DecorationOptions[]> {
```

Comment:
Part of the idea of `HighlightGrid` was we could completely decouple the production of the highlighted regions from the decorations (which gave us the ability to unit test this). Do you plan on re-adding these tests for this new API?

Reply:
Maybe. It's not easy to add tests.

Reply:
We already had a bunch of them though, and with this restructure, they're all gone. I don't mind changing up the API, but removing the tests entirely removes a lot of the safety around changing this algorithm